thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js

```diff
@@ -0,0 +1,303 @@
+      // console.log('_entitiesLoader: metricEntities', metricEntities);
+      //
+      const entities = {};
+      eventEntities.forEach(e => entities[e.urn] = e);
+      dimensionEntities.forEach(e => entities[e.urn] = e);
+      metricEntities.forEach(e => entities[e.urn] = e);
+
+      console.log('_entitiesLoader: merging cache');
+      return this._updateEntitiesCache(entities);
+    }
+  ),
+
+  _updateEntitiesCache(incoming) {
+    console.log('_updateEntitiesCache()');
+    const cache = this.get('_entitiesCache') || {};
+    Object.keys(incoming).forEach(urn => cache[urn] = incoming[urn]);
+    this.set('_entitiesCache', cache);
+    return cache;
+  },
+
+  _updateTimeseriesCache(incoming) {
+    console.log('_updateTimeseriesCache()');
+    const cache = this.get('_timeseriesCache') || {};
+    Object.keys(incoming).forEach(urn => cache[urn] = incoming[urn]);
+    this.set('_timeseriesCache', cache);
+    return cache;
+  },
+
+  _startRequestMissingTimeseries(urns) {
+    console.log('_startRequestMissingTimeseries()');
+    const pending = this.get('_pendingRequests') || new Set();
+    const cache = this.get('_timeseriesCache') || {};
+
+    const missing = new Set(urns);
+    Object.keys(pending).forEach(missing.delete);
+    Object.keys(cache).forEach(missing.delete);
+
+    console.log('_startRequestMissingTimeseries: missing', missing);
+    if (missing.size <= 0) {
+      return cache;
+    }
+
+    const metricIds = [...missing].map(urn => urn.split(":")[2]);
+
+    // NOTE: potential race condition?
+    missing.forEach(urn => pending.add(urn));
+    this.set('_pendingRequests', pending);
+
+    const idString = metricIds.join(',');
+    const analysisRange = this.get('model.analysisRange');
+
+    const url = `/timeseries/query?metricIds=${idString}&ranges=${analysisRange[0]}:${analysisRange[1]}&granularity=15_MINUTES&transformations=timestamp,relative`;
+
+    fetch(url)
+      .then(res => res.json())
+      .then(extractTimeseries)
+      .then(incoming => this._completeRequestMissingTimeseries(this, incoming));
+
+    return cache; // return current state, without new metrics
+  },
+
+  _completeRequestMissingTimeseries(that, incoming) {
+    console.log('_completeRequestMissingTimeseries()');
+    const pending = that.get('_pendingRequests') || new Set();
+
+    // NOTE: potential race condition?
+    Object.keys(incoming).forEach(urn => pending.delete(urn));
+    that.set('_pendingRequests', pending);
+
+    console.log('_completeRequestMissingTimeseries: merging cache');
+    return that._updateTimeseriesCache(incoming); // return new state, including new metrics
+  },
+
+  actions: {
+    addSelectedUrns(urns) {
+      console.log('addSelectedUrns()');
+      const selectedUrns = this.get('selectedUrns');
+      makeIterable(urns).forEach(urn => selectedUrns.add(urn));
+      this.set('selectedUrns', selectedUrns);
+      this.notifyPropertyChange('selectedUrns');
```

Comment:
do we need to specifically `notifyPropertyChange` here? (applies to other actions)

Reply:
yes. the set object reference doesn't change. we could also instantiate a new set instance instead